client/analytics/settings/index.js

```diff
@@ -158,113 +112,56 @@ class Settings extends Component {
 		window.wpNavMenuUrlUpdate( query );
 	};
 
-	handleInputChange( e ) {
+	const handleInputChange = e => {
 		const { checked, name, type, value } = e.target;
-		const { settings } = this.state;
+		const nextSettings = { ...wcAdminSettings };
 
 		if ( 'checkbox' === type ) {
 			if ( checked ) {
-				settings[ name ].push( value );
+				nextSettings[ name ].push( value );
 			} else {
-				remove( settings[ name ], v => v === value );
+				remove( nextSettings[ name ], v => v === value );
```

Comment:
This mutates the original array that is found on the existing settings object _even though_ you cast the initial object to a new object (the values of the properties will still be the original values. So for the first block, you should probably do something like this:

```
nextSettings[ name ] = [ ...nextSettings[name], value ];
```

For the second block in the condition `remove` mutates the array, so you should use the native `Array.prototype.filter` function which returns a new array:

```js
nextSettings[name] = nextSettings[name].filter( ( v ) => v !== value )
```

This also has the advantage of completely removing lodash dependency here.

It's _very important_ to be watchful against any direct mutations of state because that can lead to unexpected, and sometimes hard to troubleshoot bugs. It's something that's easy to miss in multi-dimensional arrays/objects because of shallow cloning.

Reply:
Nice catch Darren, that shallow cloning is indeed tricky.